### PR TITLE
[WIP] Refactoring code to work smoothly with multi bit sort

### PR DESCRIPTION
**Pull request by richajaindce**
Now, convert_all_bits returns vector of 2D vectors where each 2D vector is of size [ record_count ; num_multi_bits].
This helps in code clarity and is also required for malicious sort to do upgrade

### CODE REVIEW

**Comment by richajaindce** [comment]
This is a rebase on top of #409 and #412 since #409 is already approved and hopefully will be merged soon and 412 is quite straightforward. 
Once this PR is landed, this will unblock malicious sort to use multi bit permutation

File: src/protocol/attribution/aggregate_credit.rs
Comment (akoshelev) [inline src/protocol/attribution/aggregate_credit.rs:283]: is it possible to avoid allocating a vector here? would `&[x.aggregation_bit]` serve the same purpose?

File: src/protocol/modulus_conversion/mod.rs
Comment (akoshelev) [inline src/protocol/modulus_conversion/mod.rs:37]: same here, it feels like you should be able to borrow from `input` to get rid of the most inner vector. Is it possible to return an iterator here and cut the memory footprint to a constant value? this nested structure looks frightening when I think about 1M items input

File: src/protocol/modulus_conversion/mod.rs
Comment (akoshelev) [inline src/protocol/modulus_conversion/mod.rs:74]: that does not seem like a correct description

File: src/protocol/modulus_conversion/mod.rs
Comment (akoshelev) [inline src/protocol/modulus_conversion/mod.rs:81]: same idea - see if it is possible to return an iterator here

File: src/protocol/modulus_conversion/mod.rs
Comment (akoshelev) [inline src/protocol/modulus_conversion/mod.rs:88]: this makes `output` size  `record_count * input.len()` while you only fill the first `record_count` rows. Is it correct?

**Review by akoshelev — CHANGES_REQUESTED** [review]

File: src/protocol/modulus_conversion/mod.rs
Comment (richajaindce) [reply src/protocol/modulus_conversion/mod.rs:74]: This indeed is the right description. we are basically collapsing same record bits together in the same row

File: src/protocol/modulus_conversion/convert_shares.rs
Comment (richajaindce) [inline src/protocol/modulus_conversion/convert_shares.rs:171]: Doesnt compile saying "cannot move out of `locally_converted_bits`, a captured variable in an `FnMut` closure
move occurs because `locally_converted_bits` has type `I`, which does not implement the `Copy` trait"

File: src/protocol/modulus_conversion/mod.rs
Comment (akoshelev) [reply src/protocol/modulus_conversion/mod.rs:74]: I meant that you don't return `Result` so there are no errors that can be reported from `combine_slice`

File: src/protocol/modulus_conversion/convert_shares.rs
Comment (akoshelev) [reply src/protocol/modulus_conversion/convert_shares.rs]: I pushed a small change to your branch that makes this code compile. basically it did not work because you wanted to consume this iterator multiple times within `map` function and Rust wouldn't allow that. [`by_ref`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.by_ref) lets you exclusively borrow this iterator multiple times and it should work in this case, given that you own `locally_converted_bits`

Comment (richajaindce) [reply src/protocol/modulus_conversion/convert_shares.rs]: The code compiles with this change but the iterator is empty after first iteration.

File: src/protocol/modulus_conversion/mod.rs
Comment (richajaindce) [reply src/protocol/modulus_conversion/mod.rs:74]: Oh my bad.. fixing this

File: src/protocol/modulus_conversion/convert_shares.rs
Comment (richajaindce) [reply src/protocol/modulus_conversion/convert_shares.rs]: Made it work

File: src/protocol/modulus_conversion/mod.rs
Comment (richajaindce) [reply src/protocol/modulus_conversion/mod.rs]: Fixed it

File: src/protocol/modulus_conversion/mod.rs
Comment (richajaindce) [reply src/protocol/modulus_conversion/mod.rs]: This can return an iterator but the consumer of output (sort) still needs a slice. In one of the places where sort is called, we need to use the input of the function in next step: https://github.com/private-attribution/ipa/blob/main/src/protocol/ipa/mod.rs#L156

Can we keep working on the iterators in the next PR?

File: src/protocol/attribution/aggregate_credit.rs
Comment (richajaindce) [reply src/protocol/attribution/aggregate_credit.rs:287]: This needs another bunch of refactoring since this is an input to sort which is called from multiple places. Lets revisit this in the next PR?

**Comment by richajaindce** [comment]
I am still trying to see how much of the code can be iter-i-fied. But in the meantime, can we work towards landing this code and handle iterators in follow up PRs?

**Comment by richajaindce** [comment]
This looks a bit complex so making this a draft to keep discussing making the iterator change

File: src/protocol/attribution/aggregate_credit.rs
Comment (benjaminsavage) [inline src/protocol/attribution/aggregate_credit.rs:260]: I think @taikiy is/has removed this.

File: src/protocol/modulus_conversion/convert_shares.rs
Comment (benjaminsavage) [inline src/protocol/modulus_conversion/convert_shares.rs:100]: This seems highly duplicative of `convert_bit_local`. Why not invoke that?

File: src/protocol/modulus_conversion/convert_shares.rs
Comment (benjaminsavage) [inline src/protocol/modulus_conversion/convert_shares.rs:177]: That cast tho... why is it so complicated looking?

Comment (benjaminsavage) [inline src/protocol/modulus_conversion/convert_shares.rs:177]: I think you can do this in a much more idiomatic way with `.chunks(num_multi_bits)`.

https://docs.rs/itertools/0.10.5/itertools/trait.Itertools.html#method.chunks

File: src/protocol/modulus_conversion/convert_shares.rs
Comment (benjaminsavage) [inline src/protocol/modulus_conversion/convert_shares.rs:167]: I'll admit - I have no idea where this equation came from or what it does.

File: src/protocol/modulus_conversion/convert_shares.rs
Comment (benjaminsavage) [inline src/protocol/modulus_conversion/convert_shares.rs:184]: I don't think we want to call "convert_bit_list" on ALL of the bits at once. When it comes time to upgrade to malicious security, we will want to just convert the bits that are inputs to the next stage of the sort algorithm, use them, call validate, and then we can call "convert_bit_list" on the next batch, as they are inputs to the next stage, and need to be validated along with the circuit where they are used.

File: src/protocol/attribution/aggregate_credit.rs
Comment (taikiy) [reply src/protocol/attribution/aggregate_credit.rs:260]: #418. Not merged yet but it will be removed.

File: src/protocol/modulus_conversion/convert_shares.rs
Comment (martinthomson) [reply src/protocol/modulus_conversion/convert_shares.rs:167]: This rounds `num_bits` up to the next multiple of `num_multi_bits`.  I've not looked at why...

File: src/protocol/modulus_conversion/convert_shares.rs
Comment (richajaindce) [reply src/protocol/modulus_conversion/convert_shares.rs:177]: because rust doesnt let me do slice_num.try_into().unwrap() when multiplying usize numbers :(

Comment (martinthomson) [reply src/protocol/modulus_conversion/convert_shares.rs:177]: `u32::try_from(slice_num).unwrap()` would be easier to read.

**Comment by richajaindce** [comment]
Will work on iterators in future PRs

**Pull request closed without merging**
